Lineinfile module: add a new option "move" and tests for it and also "firstmatch" option

ppanczyk opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This PR adds a new option to the "lineinfile" module. I've also changed the names of options "insertbefore" and "insertafter" to "before" and "after", to be more suitable for current module capabilities. Old option names still work.
Originally, this PR introduced also a new "match_mode" option, however the same thing was introduced in  #33825 as "firstmatch".

The"move" option tells the module to move the line before or after line matching regexp specified in "before" or "after" option. It enables users to control the position of the line, even if it already exists in the specified file. To be more precise, the line is moved (and possibly replaced) when it is after the line matching "before", or before "after". In other words, it does not force the line to be right before/after line matching "before"/"after", just controls its position to be generally "after" or "before".
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
files/lineinfile module

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.5.0 (lineinfile_matching_and_moving_options 325294d0c6) last updated 2018/01/19 09:33:56 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/piotrp/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/piotrp/github/ansible/lib/ansible
  executable location = /home/piotrp/github/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
The combination of "move=yes" with "firstmatch" works as follows:
- if "firstmatch" = yes, the FIRST line matching "regexp" (or just the "line" itself, if "regexp" is not specified) is moved after or before the FIRST line matching regexp specified in "after" or "before" option, respectively (if it's not already after or before, of course);
- if "firstmatch" = no, the LAST line matching "regexp" (or "line" itself) is moved after or before the LAST line matching regexp specified in "after" or "before" option.

Such approach has been chosen because moving FIRST match of "regexp" after/before LAST match of "before"/"after" or vice versa would cause problems with idempotency in case multiple matching lines before/after line matching "before"/"after".

<!--- Paste verbatim command output below, e.g. before and after your change -->
Capabilities of both new options can be tested by the task:
```
- lineinfile:
    path: /etc/resolv.conf
    firstmatch: yes
    line: 'nameserver 8.8.8.8'
    before: '^nameserver'
    move: yes
```
The contents of resolv.conf before Ansible run:
```
# System DNS configuration
search example.com
nameserver 10.0.0.100
nameserver 10.0.0.101
nameserver 10.0.0.102
nameserver 8.8.8.8
```
Ansible run with --diff and --verbose (debug callback used):
```
--- before: /etc/resolv.conf (content)
+++ after: /etc/resolv.conf (content)
@@ -1,6 +1,6 @@
 # System DNS configuration
 search example.com
+nameserver 8.8.8.8
 nameserver 10.0.0.100
 nameserver 10.0.0.101
 nameserver 10.0.0.102
-nameserver 8.8.8.8

changed: [testserver] => {
    "backup": "", 
    "changed": true
}

MSG:

line moved
```
I have read the [module checklist](http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html) and I believe my modifications meet these requirements (except for return values documentation, but this module currently doesn't have this at all).

ppanczyk committed:
Lineinfile module: add a new option move

ppanczyk commented:
I've just fixed the typo in comment and the tests failed with timeout? Can anybody look at that? I think it's kind of a bug...

jctanner commented:
@ppanczyk we noticed from #20614 that there are no integration tests to assert idempotency in the module. Could you add a test for that and rebase your PR?

ppanczyk committed:
Tidy up the order of test tasks for lineinfile

Tidy up the order of test tasks for lineinfile module so that the file is more readable

ppanczyk committed:
Add tests for idempotency for lineinfile module

ppanczyk committed:
Add tests for new options in lineinfile module

Add tests for options firstmatch and move in lineinfile module

ppanczyk commented:
@jctanner I've done it. First I tidied up the file with tests a little bit, because it was not always matching convention described [here](http://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html#conventions) and was difficult to read. I also added some tests for the new options proposed in this PR.

ppanczyk commented:
I see that similar work was done in #33825. I've rebased this PR to combine my work with the already merged option "firstmatch". So now this adds only the "move" functionality and tests for both my new option and the "firstmatch" one.

Are there any chances for merging it?

bcoca commented:
Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we are going to decline this PR because we think this over complicates the already confusing lineinfile options and have seen no demand for this type of feature.

However, we're absolutely always up for discussion.  
Since this is a really busy project, we don't always see comments on closed tickets, but want to encourageopen dialog.  

You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

   * https://groups.google.com/forum/#!forum/ansible-devel

In the future, sometimes starting a discussion on the development list prior to implenting a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!
